### ISSUE TITLE
Mapping for all symbols defined in the C and C++ standard

### ISSUE THREAD

**Issue by amurzeau** [include-what-you-use/include-what-you-use#785]
Hi,

I'm wondering, why all defined symbols by standards are not mapped to their header instead of trying to map private header to public ones ?
Like for example strcmp symbol to string.h ?
I've dumped the C18 latest draft standard into a mapping file and this use 68KB.
Maybe for C++ it will be too large, but at least perfectly accurate for all compilers.

See here for what I've come up: [C18.imp.txt](https://github.com/include-what-you-use/include-what-you-use/files/4490349/C18.imp.txt)

The C standard happen to have a Library summary in annex B which list all defined functions and macros per headers. I used that to create that file.

Is this something IWYU would want for C/C++ in its files ?
I've not finished on the C++ side, but I'm working to try to retrieve symbols from the C++ latest draft in latex format from http://github.com/cplusplus/draft.

**Comment by kimgr** [comment]
Nice! I've thought about this before (alternatively, parsing http://en.cppreference.com), but never got around to building anything.

I think the reason it wasn't originally done this way is because it was easier to build mappings from header relationships than from actually locating symbol definitions (would require parsing C++ again to do it from standard library headers). Most of the original mappings were built with shell one-liners.

I wonder if we could distill this to a Python script we could put in the IWYU repo? And maybe combine with @trixirt's long dormant idea to generate IWYU built-in mappings from mapping files: https://github.com/include-what-you-use/include-what-you-use/pull/671. I've been meaning to pick this up some time.

**Comment by amurzeau** [comment]
Indeed parsing cppreference is probably a better idea, as I had to retrieve the text from the C18 PDF which isn't script friendly. But https://en.cppreference.com/w/cpp/symbol_index is interesting as it sum up all symbols.
When comparing the file I've (C18.imp) and cppreference, I found some discrepancies (2 typos and 2 missing functions) that I've fixed on cppreference side.

**Comment by kimgr** [comment]
Cool! I would've expected the stdc++ latex to be easier to parse, but if there's a big page with everything in it, that makes for a good target!

**Comment by amurzeau** [comment]
Yes I will see if what I can do, but parsing both can be a way to double check that cppreference doesn't contains errors (like typos) as in the C symbol index ^^
Also, the stdc++ latex won't have removing functions that were available in previous standards, so using cppreference is a good idea. (like `gets` in C, which isn't in C18).

**Comment by amurzeau** [comment]
Do you know how operators should be referenced in the iwyu mapping file ? Like this: `std::chrono::operator""h` ?

**Comment by amurzeau** [comment]
I've done a perl script that parse the wiki markup from the symbol index.
This works but it is fragile (less than parsing html I think) but mostly ok with some hacks.
That's not perfect and might have issues with missing symbols on cppreference or bad data (typos for examples) but it seems still quite good.

The quirks are in the parsing of the template syntax, when there are nested `{{ }}` ...
Maybe rewriting it in python, and maybe with a mediawiki parser python module could help, but that's a first shot (perl is definitively good at doing many regex).

Here is an archive of what I've so far with .imp files for C18 and CPP20.
[parse_symbolindex.zip](https://github.com/include-what-you-use/include-what-you-use/files/4495756/parse_symbolindex.zip)
I think C18 is nearly perfect as I can compare it to what I've extracted from the C18 standard itself.
In CPP20.imp there are inaccuracies, like atomic types aliases (ex: `atomic_bool`) mapped to `<memory>` instead of `<atomic>` because of the way the cppreference page is made.

The procedure to generate them is:
- Dump cppreference from https://en.cppreference.com/w/Special:Export, you need to enable template export. only pages in `c/`and `cpp/` folder and folders are required.
- Execute parse_cppreference_export.pl on it: `perl parse_cppreference_export.pl cppreference.com-20200417175209.xml`
- Execute parse_symbolindex.pl on a symbol index (C or CPP) and with `.` as wiki root dir: `perl parse_symbolindex.pl cpp/symbol\ index.txt .`

In the process, I've fixed some typos on cppreference so the generated files are ok.

**Comment by kimgr** [comment]
> Do you know how operators should be referenced in the iwyu mapping file ? Like this:
> `std::chrono::operator""h` for https://en.cppreference.com/w/cpp/chrono/operator%22%22h ?

Sorry, I missed this question.

I don't know. I think the rule is that the symbol in mappings should be on the same form as the symbol in IWYU output, so if you can construct an example to have IWYU report a diagnostic about that/an operator, you should get a good idea. It's not unthinkable that there may be bugs in reporting format with more exotic constructs, so if it looks crazy maybe we should address that first.

**Comment by kimgr** [comment]
I'm not having any luck with this:

> Dump cppreference from https://en.cppreference.com/w/Special:Export, you need to enable
> template export. only pages in c/and cpp/ folder and folders are required.

Can you expand a little on what the exact strings to put in the multiline textbox are? I keep getting an empty-looking export. Thanks!

Overall, it seems to me this is a good way forward, and the scripts look simple-ish enough to be easy to port to Python. If we could build something small to bootstrap the export from the script as well, this would be really ergonomic.

**Comment by amurzeau** [comment]
> Can you expand a little on what the exact strings to put in the multiline textbox are? I keep getting an empty-looking export. Thanks!

On the export page, you need to:

- Ensure `Add pages from namespace: ` is set to `Main` and click add
- This will add all pages to the textbox bellow
- To reduce the export size (and maybe cppreference' server resources usage), remove everything under `old/` in the textbox
- Check the 3 checkbox at the bottom of the page:
  - Include only the current revision, not the full history
  - Include templates
  - Save as file
- Click on the `export` button, this should ask you where to save the `cppreference.com-<timestamp>.xml` file

> 
> Overall, it seems to me this is a good way forward, and the scripts look simple-ish enough to be easy to port to Python. If we could build something small to bootstrap the export from the script as well, this would be really ergonomic.

I've tried to retrieve the curl command from firefox itself and this generate the POST request with its data. But it might be not good enough, as the exported pages are hardcoded into the curl request.

This issue is that when clicking on the `Add` button to populate the textbox, it won't use a webservice, but just reload the whole page with the textbox fulfilled.

**Comment by amurzeau** [comment]
I've tried to parse cplusplus/draft sources in latex, but I will give up for now, that requires too much parsing code again to handle braces and namespace + other things.
I'm putting the perl script here just in case but won't work more on it (for now at least).
[parse_cplusplusdraft.pl.gz](https://github.com/include-what-you-use/include-what-you-use/files/4500577/parse_cplusplusdraft.pl.gz)

**Comment by amurzeau** [comment]
Note: having the symbol mapping still require some header mapping, as internal stuff is leaking like this:
```cpp
#include <xhash>                       // for _Hash<>::iterator
#include <xstring>                     // for operator==, operator<
#include <xtree>                       // for _Tree_iterator, _Tree<>::iterator
```

**Comment by kimgr** [comment]
@amurzeau Thanks for the added detail, I haven't had a chance to play with that yet, but I will.

As for the internal stuff leaking out, I think some of that is IWYU picking the wrong symbol for uses (with typedefs in templates, for example). Those are bugs, and I don't think we should use header mappings to work around them. Well, maybe we should, to improve the user experience, but it needs to be carefully tracked so we don't mask bugs with a heavy-handed, non-portable mechanism like private-public header mappings like we have up until now.

However, the other example you cite is interesting: 

    #include <xstring>                     // for operator==, operator<

It's hard to put in symbol mappings for operator overloads, as the spelling IWYU spits out doesn't mention the parameter types (`std::string` and probably `const char*` in this case).

So maybe for such operators, private-public header mappings are the right tool?

**Comment by amurzeau** [comment]
Yes I guess header mapping is still a good thing, and it's convenient (there are less headers to map than symbols).

**Comment by bolshakov-a** [comment]
> It's hard to put in symbol mappings for operator overloads

I'm wondering: should IWYU ever report such standard library operators? Because if you compare two `std::string`s like `str1 == str2`, you should have `std::string` explicitly mentioned in your _translation unit_ (either at the variable declarations, or at the declaration of `GetString()` if you declare `str1` like this: `auto str1 = GetString();`), right? And because the C++ standard doesn't allow forward-declarations of names from `std` in user's code, you should already have `#include <string>` somewhere in the TU.

So maybe, `std::operator==` should be mapped (explicitly) to nothing? (Just conceptually. I'm not sure if this currently would work in IWYU).

**Comment by kimgr** [comment]
@bolshakov-a That makes sense to me. Could probably be a rule in IWYU proper to ignore operators in `std`.

I'm only thinking about stuff like:

* Is it possible to take the address of `std::operator==(const std::string&, const std::string&)`? I guess in order to do so, we would have to mention `std::string` anyway? Or is it even possible to do?
* It seem unlikely, but possible, to have some operator for a type `std::foo` declared in a header separate from `foo` itself. It seems like a really contrived design, however.

**Comment by bolshakov-a** [comment]
> Is it possible to take the address of std::operator==(const std::string&, const std::string&)? I guess in order to do so, we would have to mention std::string anyway? Or is it even possible to do?

Since C++20, there is only [a limited set of addressable functions](https://www.en.cppreference.com/w/cpp/language/extending_std.html#Addressing_restriction) in the standard library. I think this applies to operators as well, so they are not addressable anymore.

> It seem unlikely, but possible, to have some operator for a type std::foo declared in a header separate from foo itself.

Yeah, the mapping should be made carefully.

> Could probably be a rule in IWYU proper to ignore operators in `std`.

So I think it should be explicit mapping, not a generic rule, because e.g. `operator""sv` should still trigger a suggestion of `#include <string_view>`, but some functions like `std::swap` might be ignored as well.

**Comment by kimgr** [comment]
OK, that makes sense. So something like:
```
symbol: ["std::foobar", "private", "", "public"]
```
?

i.e. use empty string as a sentinel meaning "drop the symbol use"?

**Comment by bolshakov-a** [comment]
Yeah, like this. I think the rule should be as follows: if a function takes at least one class from `std`, and is located in the same header as the class, and is not addressable, then it should be mapped to `""`. I don't think the should support possible (?) fragile cases allowed by pre-C++20 standards.

**Comment by kimgr** [comment]
Sounds good to me!

**Comment by bolshakov-a** [comment]
One concern about mapping to nothing. User's code should not contain forward-declarations of names from `std`, but a standard library implementation code may do this. What if some standard library function returns a type defined in other standard library header? Should the header defining the function necessarily include the header defining the type?

As an example, `<stacktrace>` defines `to_string` function returning `std::string`, as well as `stacktrace_entry` class has a couple of methods returning it. However, I see nothing in the standard indicating that `<stacktrace>` should include `<string>`, despite libstdc++ and Microsoft implementations do that (libc++ does not implement `<stacktrace>` yet). So, if a user compares results of two `to_string` calls, should IWYU suggest `<string>` header for the comparison operator? Currently, IWYU should report `std::string` anyway because it is returned by value and not e.g. by reference. But I planned to suppress reporting of unwritten names from `std` by the same reasons as described above.

**Comment by kimgr** [comment]
OK, so in user code we work with the author-intent hint for function return type:
```
// Forward-declare as a signal that caller is responsible for including.
// Otherwise the callee is responsible.
struct RetType;
RetType Function();
```

Do I understand it right that your describing the same problem, but for standard library code?

I don't have a clear view of the entire problem top-to-bottom, so I have a hard time forming opinions. Could you show a small example with `stacktrace` and its `to_string` and comparisons, and current/desired IWYU diagnostics to maybe help me catch on? Thanks.

**Comment by bolshakov-a** [comment]
Usually if you have your own header with `GetString()` function:
```cpp
// get_string.h
#include <string>

std::string GetString(int);
```
and then use it like this:
```cpp
// user.cpp
#include "get_string.h"

bool User() {
  auto str1 = GetString(1);
  auto str2 = GetString(2);
  return str1 == str2;
}
```
there is no need to include `<string>` in `user.cpp` explicitly: `get_string.h` should already provide it because you should not forward-declare `std` stuff. (Currently, IWYU suggests to include `<string>`, but I think it is redundant.) However, I'm not sure if such logic applies to standard library headers, like in this case:
```cpp
#include <stacktrace>

bool Test() {
  auto str1 = to_string(std::stacktrace::current());
  auto str2 = to_string(std::stacktrace::current());
  return str1 == str2;
}
```
Is it guaranteed to compile, or `<string>` should be included explicitly for portability? Currently, again, IWYU suggests `<string>`. But I'm not sure what the desired behavior is. Maybe, we should ask on StackOverflow or somewhere else...

**Comment by bolshakov-a** [comment]
Looks like mapping to nothing is a bad idea, at least for functions taking `std::string` as a parameter: https://discourse.llvm.org/t/should-stacktrace-provide-string/89605
